packages/xdl/src/detach/AndroidShellApp.js

```diff
@@ -1111,3 +1114,102 @@ export function addDetachedConfigToExp(exp: Object, context: StandaloneContext):
   );
   return exp;
 }
+
+/**
+
+Some files in `android` directory have the following patterns of code:
+
+```
+// WHEN_PREPARING_SHELL_REMOVE_FROM_HERE
+
+// BEGIN_SDK_30
+some SDK 30-specific code
+// END_SDK_30
+
+// BEGIN_SDK_29
+some SDK 29-specific code
+// END_SDK_29
+
+...
+
+// WHEN_PREPARING_SHELL_REMOVE_TO_HERE
+```
+
+The following function replaces `BEGIN_SDK_XX` with `REMOVE_TO_HERE`
+and `END_SDK_XX` with `REMOVE_FROM_HERE`.
+
+This allows us to use `deleteLinesInFileAsync` function to remove obsolete SDKs code easily.
+
+ */
+const removeInvalidSdkLinesWhenPreparingShell = async (majorSdkVersion, filePath) => {
+  await regexFileAsync(
+    `BEGIN_SDK_${majorSdkVersion}`,
+    `WHEN_PREPARING_SHELL_REMOVE_TO_HERE`,
+    filePath
+  );
+  await regexFileAsync(
+    `END_SDK_${majorSdkVersion}`,
+    `WHEN_PREPARING_SHELL_REMOVE_FROM_HERE`,
+    filePath
+  );
+  await deleteLinesInFileAsync(
+    'WHEN_PREPARING_SHELL_REMOVE_FROM_HERE',
+    'WHEN_PREPARING_SHELL_REMOVE_TO_HERE',
+    filePath
+  );
+};
+
+async function removeObsoleteSdks(shellPath: string, requiredSdkVersion: string) {
+  const majorSdkVersion = semver.major(requiredSdkVersion);
+
+  // Remove obsolete `expoview-abiXX_X_X` dependencies
+  const appBuildGradle = path.join(shellPath, 'app', 'build.gradle');
+  await removeInvalidSdkLinesWhenPreparingShell(majorSdkVersion, appBuildGradle);
+
+  // Remove obsolete `host.exp.exponent:reactandroid:XX.X.X` dependencies from expoview
+  const expoviewBuildGradle = path.join(shellPath, 'expoview', 'build.gradle');
+  await removeInvalidSdkLinesWhenPreparingShell(majorSdkVersion, expoviewBuildGradle);
+
+  // Remove no-longer-valid interfaces from MultipleVersionReactNativeActivity
+  const multipleVersionReactNativeActivityJava = path.join(
+    shellPath,
+    'expoview',
+    'src',
+    'main',
+    'java',
+    'host',
+    'exp',
+    'exponent',
+    'experience',
+    'MultipleVersionReactNativeActivity.java'
+  );
```

Comment:
Lol, why so many parameters when you can just do:
```'expoview/src/main/java/host/exp/exponent/experience/MultipleVersionReactNativeActivity.java'```, i think it doesn't work on Windows anyway.

Reply:
Hmm, I believe the Android scripts are meant to work on Windows 🤔 is it a known issue that they do not?

Reply:
Actually, I don't know if it's a known issue, but if there is any OS on which these scripts don't work it's probably Windows :D I meant that I don't know if there are people who use these scripts on Windows.
Generally, I thought that using `/` to separate directories in `path.join` calls could be a problem on Windows, but I just checked that even if the directories are separated by `/` and the script is executed on Windows, node changes them to `\\`. So everything should work as expected and we don't have any issue here.